CO-2269 Dissocier la communication Welcome Muskathlon d'un nouveau Portal User

jo-bloiz-ornot opened the pull request:
Wizard that will let you choose if you want to create a communication and let you choose which template (for now there is only Muskathlon template)

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,60 @@
+        # you should give context
+        # { 'create_uid_communication': False }
+        # else it will create a communication
+        if self.env.context['create_communication_with_uid'] is None:
```
Instead of reading this value from the context, it's better and more clear to directly add a field in the wizard 'portal.wizard.user' model. `invitation_config_id = fields.Many2one('partner.communication.config')`. With that you can here directly test `if self.invitation_communication_id:`

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+
+    @api.multi
+    def button_create_portal_user(self):
+        portal = self.env['portal.wizard'].create({})
```
If you add the field `invitation_config_id` in the 'portal.wizard.user' model, you can here pass it in the creation values `{'invitation_config_id': self.config_id}`

jo-bloiz-ornot replied:
The problem is that it's not a portal wizard user but just a portal wizard, I think overriding the method (onchange_id) that create portal user is possible but it's will lose the adaptability of the oca module

jo-bloiz-ornot replied:
I think I can do it without losing changes of oca module, I tell you later

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,60 @@
+            raise UserError(_('You must have an email address in'
+                              ' your User Preferences to send emails.'))
+
+        config_id_id = self.env.context['config_id_id'] or \
```
If you add the field, here the communication would be simply `self.invitation_config_id`. I don't think that you should catch a case where the user wants to send something but doesn't select any communication rule. It makes no sense, either they choose one to send or we don't send nothing.

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+    _description = 'take a partner and make it a odoo user'
+
+    send_communication = fields.Boolean(
+        _('do you want to send a mail with '
```
Field strings don't need to be translated with the python function, they are automatically included in translations. Try to keep them short or it will become too long in the user interface. Here I propose simply `send_communication = fields.Boolean('Send an e-mail invitation')`

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+          'the login link to the user ?'), default=True)
+
+    config_id = fields.Many2one(
+        'partner.communication.config', _('choose a communication config'),
```
'Choose a communication'

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+        bool_communication_with_uid = \
+            (self.send_communication and self.config_id)
+
+        ctx = {
```
No need to put a custom context if you pass the config via the field

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+    @api.multi
+    def button_create_portal_user(self):
+        portal = self.env['portal.wizard'].create({})
+        portal.onchange_portal_id()
```
Add a comment here to explain that the following lines are for allowing registering users even if they don't have an e-mail address.

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,60 @@
+                'config_id': config_id_id
+            })
+
+        return True
```
Instead of returning True, you can return the communication created in order to use it directly. `return self.env['partner.communication.job'].create`

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,60 @@
+                self.env.context['create_communication_with_uid']
+
+        if bool_communication_with_uid:
+            self._create_uid_communication()
```
Here you can add inside res the communication created (or just the id maybe if it doesn't work). `res['communication'] = self._create_uid_communication()`. `_create_uid_communication` has of course to return the communication record.

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+        action = None
+        if bool_communication_with_uid:
+            # get same communication create in portal.action_apply()
+            same_job_search = [
```
If your function `action_apply` puts the communication inside the result, here you can simply retrieve it by doing `communication = action_result['communication']`

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,80 @@
+            'create_communication_with_uid': bool_communication_with_uid,
+            'config_id_id': self.config_id.id
+        }
+        portal.with_context(ctx).action_apply()
```
assign the result in order to retrieve the generated communication : `action_result = portal.action_apply()`

ecino reviewed (changes requested):
Very good changes, thank you! Here are some suggestions in order to improve how parameters and results are passed between the different objects and methods.

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,73 @@
-                'config_id': config.id
-            })
-        return res
+        self.ensure_one()
```
You removed the call to the super which here can remove some useful features. Can you add it back?

jo-bloiz-ornot replied:
No, because the parent didn't return the result of the action_apply method

```
    @api.multi
    def action_apply(self):
        self.ensure_one()
        self.user_ids.action_apply()
        return {'type': 'ir.actions.act_window_close'}
```
Override had to be done there

ecino replied:
The problem is that this code will never be executed :
https://github.com/compassionch/compassion-modules/blob/bac7e809eb6deabea16c080698f84122601a69d9/crm_compassion/wizards/portal_wizard.py#L21
You should call the super even if it doesn't return what we need and then return what you need after. Maybe here some conception needs to be changed to ensure we don't lose the code of crm_compassion.

jo-bloiz-ornot replied:
ok

Just one question, is the fact that the method portal.wizard.user action_apply() will be call once in the super method then again just to get the result is a problem ?

jo-bloiz-ornot replied:
I will try with a additional field in portal.wizard.user to not call twice the method

jo-bloiz-ornot replied:
it's a success

jo-bloiz-ornot committed:
CO-2269 Dissocier la communication Welcome Muskathlon d'un nouveau Portal User

Wizard that will let you choose if you want to create a communication and let you choose which template (for now there is only Muskathlon template)

Correction pour travis

move the logic to partner_communication_switzerland

Correction de travis

Requested change

change vars name

for portal and portal.user
action_apply now return a communication

use list to form for showing result

change how value is return to use the super method again

jo-bloiz-ornot committed:
correction for travis

jo-bloiz-ornot committed:
correction for travis

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,65 @@
+
+        # if the context is not given, create a communication
+        # else do as context say
+        if self.env.context['create_communication'] is None or \
```
I don't understand why you create a communication when context is None. To me, it sounds more logic to create the communication only when we are told to do so. This would mean `if self.env.context.get('create_communication'):`

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,71 @@
+
+        no_mail.mapped('partner_id').write({'email': False})
+
+        action = None
```
`action = True` is better to return.

ecino commented on partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:
```diff
@@ -0,0 +1,71 @@
+
+    @api.multi
+    def button_cancel(self):
+        return None
```
return True (means ok).

jo-bloiz-ornot committed:
Requested change + add automatic account activation on no email sending

jo-bloiz-ornot committed:
correction making login for partner

request change

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,71 @@
+
+        self.mapped('partner_id').signup_prepare()
+
+        if self.env.context['create_communication']:
```
```suggestion
        if self.env.context.get('create_communication'):
```
This avoids the code execution to fail in case `create_communication` is not present in the context

jo-bloiz-ornot replied:
good to know, thank you

ecino commented on partner_communication_switzerland/wizards/portal_wizard.py:
```diff
@@ -9,25 +9,71 @@
+            for wizard_line in self:
+                wizard_line.create_uid_communication()
+        else:
+            for wizard_line in self:
```
Why did you add this? I'm not sure we want to automatically activate the account if we don't send a communication.

jo-bloiz-ornot replied:
I was afraid to give to much work to the user to activate manually the account, because if you dont send a email with a link to activate the account, who gonna activate it ?

ecino replied:
In fact we register some users that we never want to give access to the account, only to be able to assign them to some events or other objects (but this is intern stuff and we don't communicate to them). So we don't wan't that everyone is activated automatically. I think you can remove this part.

jo-bloiz-ornot replied:
alright, it's gonna be done
